main.go

```diff
@@ -119,11 +115,56 @@ func processRequests(deathNote map[string]bool, firstConnected chan<- bool, wg *
 					break
 				}
 			}
-			log.Printf("Client disconnected: %s\n", conn.RemoteAddr().String())
-			conn.Close()
+		}(conn)
+	}
+}
+
+func waitForPruneCondition(connectionAccepted <-chan net.Addr, connectionLost <-chan net.Addr) {
+	connectionCount := 0
+	never := make(chan time.Time, 1)
+	defer close(never)
 
-			time.Sleep(10 * time.Second)
-		}()
+	handleConnectionAccepted := func(addr net.Addr) {
+		log.Printf("New client connected: %s", addr)
+		connectionCount++
+	}
+
+	select {
+	case <-time.After(initialConnectTimeout):
+		panic("Timed out waiting for the first connection")
+	case addr := <-connectionAccepted:
+		handleConnectionAccepted(addr)
+	}
+
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
```

Comment:
what bothers me, that we listen for signals in a function, not in a main. Also as we refactor that anyways - maybe there is a chance to always handle cleanup as a context close with NotifyContext (even during the initial timeout).
I will try to explain what I think with a code.

Reply:
There might be trade-offs if there signals are registered earlier, but probably depends on the implementation. Before `signal.Notify` it is still possible to just stop the program via those signal. Only after initialization it is needed to handle those signals explicitly and also not call `signal.Stop` again, as this would make it possible to abort Ryuk while it is trying to cleanup. So I am not sure where you want to cleanup.

Reply:
WDYT of this?

```diff
Index: main.go
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main.go b/main.go
--- a/main.go	(revision 9bb57be7dd79b35ad1adb24abc015a161a9d7a95)
+++ b/main.go	(date 1643805325365)
@@ -8,7 +8,6 @@
 	"log"
 	"net"
 	"net/url"
-	"os"
 	"os/signal"
 	"strings"
 	"syscall"
@@ -50,8 +49,10 @@
 	connectionLost := make(chan net.Addr)
 
 	go processRequests(deathNote, connectionAccepted, connectionLost)
+	ctx, stop := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	defer stop()
 
-	waitForPruneCondition(connectionAccepted, connectionLost)
+	waitForPruneCondition(ctx, connectionAccepted, connectionLost)
 
 	prune(cli, deathNote)
 }
@@ -119,7 +120,7 @@
 	}
 }
 
-func waitForPruneCondition(connectionAccepted <-chan net.Addr, connectionLost <-chan net.Addr) {
+func waitForPruneCondition(ctx context.Context, connectionAccepted <-chan net.Addr, connectionLost <-chan net.Addr) {
 	connectionCount := 0
 	never := make(chan time.Time, 1)
 	defer close(never)
@@ -134,13 +135,10 @@
 		panic("Timed out waiting for the first connection")
 	case addr := <-connectionAccepted:
 		handleConnectionAccepted(addr)
+	case <-ctx.Done():
+		log.Println("Signal received")
+		return
 	}
-
-	signals := make(chan os.Signal, 1)
-	signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
-	// Do not stop handling the signals on returning from this function
-	// as we do not want the pruning process to be interrupted ungracefully.
-	// defer signal.Stop(signals)
 
 	for {
 		var noConnectionTimeout <-chan time.Time
@@ -158,8 +156,8 @@
 			log.Printf("Client disconnected: %s", addr.String())
 			connectionCount--
 			break
-		case sig := <-signals:
-			log.Println("Signal received:", sig)
+		case <-ctx.Done():
+			log.Println("Signal received")
 			return
 		case <-noConnectionTimeout:
 			log.Println("Timed out waiting for re-connection")
```

this way we don't hide the signal handling in the function and it's still explicit in the main

Reply:
@lanwen Looks much better! :+1: